x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/LinearProgrammingPlanSolver.java

```diff
@@ -0,0 +1,341 @@
+            logger.debug(() -> "Best plan is from bin packing");
+        } else {
+            logger.debug(() -> "Best plan is from LP solver");
+        }
+
+        return assignmentPlan;
+    }
+
+    private double weightForAllocationVar(Model m, Node n, Map<Tuple<Model, Node>, Double> weights) {
+        return (1 + weights.get(Tuple.tuple(m, n)) - (m.memoryBytes() > n.availableMemoryBytes() ? 10 : 0)) - L1 * normalizedMemoryPerModel
+            .get(m) / maxNodeCores;
+    }
+
+    private Tuple<Map<Tuple<Model, Node>, Double>, AssignmentPlan> calculateWeightsAndBinPackingPlan() {
+        logger.debug(() -> "Calculating weights and bin packing plan");
+
+        double w = INITIAL_W;
+        double dw = w / nodes.size() / models.size();
+
+        Map<Tuple<Model, Node>, Double> weights = new HashMap<>();
+        AssignmentPlan.Builder assignmentPlan = AssignmentPlan.builder(nodes, models);
+
+        for (Model m : models.stream().sorted(Comparator.comparingDouble(this::dsafModelOrder)).toList()) {
+            while (true) {
+                List<Node> orderedNodes = nodes.stream()
+                    .sorted(Comparator.comparingDouble(n -> dsafNodeOrder(n, m, assignmentPlan)))
+                    .toList();
+                double lastW = w;
+                for (Node n : orderedNodes) {
+                    int allocations = Math.min(
+                        assignmentPlan.getRemainingCores(n) / m.threadsPerAllocation(),
+                        assignmentPlan.getRemainingAllocations(m)
+                    );
+                    if (allocations > 0 && assignmentPlan.canAssign(m, n, allocations)) {
+                        assignmentPlan.assignModelToNode(m, n, allocations);
+                        weights.put(Tuple.tuple(m, n), w);
+                        w -= dw;
+                        break;
+                    }
+                }
+                if (lastW == w || assignmentPlan.getRemainingAllocations(m) == 0) {
+                    break;
+                }
+            }
+        }
+
+        for (Model m : models) {
+            for (Node n : nodes) {
+                if (weights.containsKey(Tuple.tuple(m, n)) == false) {
+                    weights.put(Tuple.tuple(m, n), random.nextDouble(minWeight(m, n, w), maxWeight(m, n, w)));
+                }
+            }
+        }
+
+        logger.trace(() -> "Weights = " + weights);
+        AssignmentPlan binPackingPlan = assignmentPlan.build();
+        logger.debug(() -> "Bin packing plan =\n" + binPackingPlan.prettyPrint());
+
+        return Tuple.tuple(weights, binPackingPlan);
+    }
+
+    private double dsafModelOrder(Model m) {
+        return (m.currentAllocationByNodeId().isEmpty() ? 1 : 2) * -normalizedMemoryPerModel.get(m) * m.threadsPerAllocation();
+    }
+
+    private double dsafNodeOrder(Node n, Model m, AssignmentPlan.Builder assignmentPlan) {
+        // as java.lang.Math#abs(long) is a forbidden API, we compute an abs manually
```

Comment:
You should do something similar `BucketCountKSTestAggregator` except with `long`

```
 @SuppressForbidden(reason = "Math#abs(int) is safe here as we protect against MIN_VALUE")
    private static int nonNegative(int x) {
        if (x == Integer.MIN_VALUE) {
            throw new AggregationExecutionException("unexpected value while interpolating sampled values");
        }
        return Math.abs(x);
    }
```